docs/tutorial/reduction.rst

```diff
@@ -0,0 +1,728 @@
+
+.. note::
+
+    This is important, because on a GPU there is no stack, but local memory is provisioned from the register file and this provisioning happens statically. To paraphrase, the address range of a thread's local memory is determined at compile time. When an array is defined and used in local storage, the compiler can only maintain its storage in the register file as long as all access to the array is computable by the compiler at compile-time. It need not strictly be a compile-time constant, if through constant folding or some other means the compiler can resolve the addresses of the accesses. However, if it cannot, the array will be backed by global memory (indicated by allocating a non-zero number of spill registers observable using static analysis tools) which is multiple orders of magnitude slower. ``hip::static_array`` via its ``hip::get<>`` interface guarantees that no such spills will occur.
+
+.. code-block:: C++
+
+    template<uint32_t BlockSize, uint32_t WarpSize, uint32_t ItemsPerThread>
+    __global__ static __launch_bounds__(BlockSize) void kernel(...)
+
+Our kernel as promised now has three compile-time configurable parameters. The only part of the kernel that changes is how we load data from global and how we perform the binary operation on those loaded values. What used to be the one-liner:
+
+.. code-block:: C++
+
+    // Read input from front buffer to local
+    T res = op(read_global_safe(gid), read_global_safe(gid + blockDim.x));
+
+is going to be split now to a reading and a processing step.
+
+Reading ``ItemsPerThread``
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The change to reading is going to happen inside `read_global_safe`:
+
+.. code-block:: C++
+
+    auto read_global_safe = [&](const int32_t i) -> hip::static_array<T, ItemsPerThread>
+    {
+        return [&]<int32_t... I>(std::integer_sequence<int32_t, I...>)
+        {
+            if(i + ItemsPerThread < front_size)
+                return hip::static_array<T, ItemsPerThread>{
+                    front[i + I]...
+                };
+            else
+                return hip::static_array<T, ItemsPerThread>{
+                    (i + I < front_size ? front[i + I] : zero_elem)...
+                };
+        }(std::make_integer_sequence<int32_t, ItemsPerThread>());
+    };
+
+What's happening here? Without the flexibility of a configurable ``ItemsPerThread`` property, we'd want to load each array element one after the other, morally equivalent to:
+
+.. code-block:: C++
+
+    T arr[4] = {
+        front[gid + 0],
+        front[gid + 1],
+        front[gid + 2],
+        front[gid + 3]
+    }
+
+This is exactly what's happening in the ``front[i + I]...`` fold-expression. There is a condition though: we only issue this if the entire read is operating on real input and it's not padding using ``zero_elem``. If some reads would overindex the input, the read turns into:
+
```

Comment:
```suggestion
This is exactly what's happening in the ``front[i + I]...`` fold-expression. However, this can only be issued if the entire read is operating on real input without padding using ``zero_elem``. If some reads over-index the input, the read turns into:

```